### PR TITLE
[#276] Employee can add Issues

### PR DESCRIPTION
**Pull request by sandeepc19**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@ def current_user
   end
 
   def authenticate_user
-    redirect_to root_path unless current_user
+    redirect_to root_path unless current_user.admin?
```
Comment (nbanerjee) [inline app/controllers/application_controller.rb]: Why this change?

File: app/models/employee.rb [hunk 2]
```diff
@@ -6,5 +6,11 @@ class Employee < ActiveRecord::Base
   validates :name, presence: true
   validates :email, uniqueness: { case_sensitive: false }, format: { with: User::VALID_EMAIL_REGEX }, allow_blank: true
 
+  after_save :check_employee_email
```
Comment (nbanerjee) [inline app/models/employee.rb]: Please use proper callback

File: app/models/employee.rb [hunk 3]
```diff
@@ -6,5 +6,11 @@ class Employee < ActiveRecord::Base
...
+
   scope :order_by_name, -> { order('name') }
+
+  def check_employee_email
+    User.find_or_create_by(email: Employee.find(id).email)
```
Comment (nbanerjee) [inline app/models/employee.rb]: Why you need to find employee?

File: db/migrate/20170222061836_add_admin_to_users.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddAdminToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :admin, :boolean
```
Comment (nbanerjee) [inline db/migrate/20170222061836_add_admin_to_users.rb]: It should have default value

**Comment by nbanerjee** [comment]
When employee is inactive then corresponding user should not access this system.

Rebase your branch

File: app/models/user.rb [hunk 5]
```diff
@@ -13,6 +13,10 @@ def full_name
     "#{first_name} #{last_name}".squish
   end
 
+  def self.corresponding_user(user)
+    Employee.find_by_email(user.email)
+  end
```
Comment (nbanerjee) [inline app/models/user.rb]: Why class method? Why inconsistency in naming and code?

File: app/services/fetch_external_employee.rb [hunk 6]
```diff
@@ -4,6 +4,7 @@ def self.add_employees
       employee = Employee.find_or_create_by(name: emp.name, external_id: emp.id)
       employee.email = emp.email
       employee.save
+      User.find_or_create_by(email: emp.email)
```
Comment (nbanerjee) [inline app/services/fetch_external_employee.rb]: Why do you need this?

File: app/models/user.rb [hunk 7]
```diff
@@ -13,6 +14,10 @@ def full_name
     "#{first_name} #{last_name}".squish
   end
 
+  def employee(user)
```
Comment (nbanerjee) [inline app/models/user.rb]: Why you need `user` as parameter?

Comment (nbanerjee) [inline app/models/user.rb]: And what is the use of this function?

**Comment by nbanerjee** [comment]
When employee is getting deactivated then corresponding user also deactivated.

**Comment by nbanerjee** [comment]
Add migration for existing employees.

**Comment by nbanerjee** [comment]
Please add spec for this feature.

File: app/models/employee.rb [hunk 8]
```diff
@@ -1,16 +1,31 @@
...
+
+  def create_corresponding_user
+    user = User.find_or_create_by(email: email)
+    self.user_id = user.id
+    save
+  end
+
+  def delete_user_if_employee_inactive
+    if user.present?
+      user.destroy if active_changed? && active == false
```
Comment (nbanerjee) [inline app/models/employee.rb]: this condition should be in top, and `user.present?` should be here.

File: db/migrate/20170310114118_add_user_id_to_existing_employees.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+class AddUserIdToExistingEmployees < ActiveRecord::Migration
+  def up
+    employees = Employee.all
+    employees.each do |emp|
+      if emp.user_id.nil?
```
Comment (nbanerjee) [inline db/migrate/20170310114118_add_user_id_to_existing_employees.rb]: Find all those employees who doesn't have access to login in this application and do the task.

File: app/models/employee.rb [hunk 10]
```diff
@@ -1,16 +1,31 @@
...
+    user = User.find_or_create_by(email: email)
+    self.user_id = user.id
+    save
+  end
+
+  def delete_user_if_employee_inactive
+    if active_changed? && active == false
+      user.destroy if user.present?
+    end
+  end
```
Comment (nbanerjee) [inline app/models/employee.rb]: combined `deallocate_items` & `delete_user_if_employee_inactive` methods into one.

File: db/migrate/20170309072540_add_user_id_to_employees.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+class AddUserIdToEmployees < ActiveRecord::Migration
+  def change
+    add_column :employees, :user_id, :integer
```
Comment (nbanerjee) [inline db/migrate/20170309072540_add_user_id_to_employees.rb]: use add reference and add index.

**Comment by ssinghi** [comment]
Please create a separate set of controller and routes accessible by employees. I don't want any overloading of routes between admins and other roles.
